### PR TITLE
feat: support new reporting component

### PR DESCRIPTION
**Pull request by pasha-codefresh**

### CODE REVIEW

File: Procfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
...
-redis: bash -c "if [ \"$ARGOCD_REDIS_LOCAL\" = 'true' ]; then redis-server --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; else docker run --rm --name argocd-redis -i -p ${ARGOCD_E2E_REDIS_PORT:-6379}:${ARGOCD_E2E_REDIS_PORT:-6379} docker.io/library/redis:$(grep "image: redis" manifests/base/redis/argocd-redis-deployment.yaml | cut -d':' -f3) --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; fi"
+redis: bash -c "if [ \"$ARGOCD_REDIS_LOCAL\" = 'true' ]; then redis-server --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; else docker run --rm --name argocd-redis -i -p ${ARGOCD_E2E_REDIS_PORT:-6379}:${ARGOCD_E2E_REDIS_PORT:-6379} docker.io/library/redis:7.0.14-alpine --save '' --appendonly no --port ${ARGOCD_E2E_REDIS_PORT:-6379}; fi"
```
Comment (pasha-codefresh) [inline Procfile]: revert

File: manifests/base/event-reporter/event-reporter-statefulset.yaml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      app.kubernetes.io/name: event-reporter
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: event-reporter
+    spec:
+      serviceAccountName: event-reporter
+      containers:
+      - name: event-reporter
+        image: docker.io/xeonalex/personal-argocd-dev:reporting-v2-test-v20
```
Comment (yaroslav-codefresh) [inline manifests/base/event-reporter/event-reporter-statefulset.yaml]: need to change to release version

File: cmd/event-reporter-server/commands/event_reporter_server.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	servercache "github.com/argoproj/argo-cd/v2/server/cache"
+	"github.com/argoproj/argo-cd/v2/util/cli"
+	"github.com/argoproj/argo-cd/v2/util/env"
+	"github.com/argoproj/argo-cd/v2/util/errors"
+	"github.com/argoproj/argo-cd/v2/util/kube"
+	"github.com/argoproj/argo-cd/v2/util/tls"
+)
+
+const (
+	failureRetryCountEnv              = "ARGOCD_K8S_RETRY_COUNT"
```
Comment (pasha-codefresh) [inline cmd/event-reporter-server/commands/event_reporter_server.go]: rename

File: cmd/event-reporter-server/commands/event_reporter_server.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	failureRetryCount              = 0
+	failureRetryPeriodMilliSeconds = 100
+)
+
+func init() {
+	failureRetryCount = env.ParseNumFromEnv(failureRetryCountEnv, failureRetryCount, 0, 10)
+	failureRetryPeriodMilliSeconds = env.ParseNumFromEnv(failureRetryPeriodMilliSecondsEnv, failureRetryPeriodMilliSeconds, 0, 1000)
+}
+
+// NewCommand returns a new instance of an argocd command
```
Comment (pasha-codefresh) [inline cmd/event-reporter-server/commands/event_reporter_server.go]: rename argocd

File: cmd/event-reporter-server/commands/event_reporter_server.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+		argocdToken              string
+		codefreshUrl             string
+		codefreshToken           string
+		shardingAlgorithm        string
+	)
+	var command = &cobra.Command{
+		Use:   cliName,
+		Short: "Run the Event Reporter server",
+		// TODO: [Event Reporter] change description
+		Long:              "The API server is a gRPC/REST server which exposes the API consumed by the Web UI, CLI, and CI/CD systems.  This command runs API server in the foreground.  It can be configured by following options.",
```
Comment (pasha-codefresh) [inline cmd/event-reporter-server/commands/event_reporter_server.go]: rename description

File: common/common.go [hunk 6]
```diff
@@ -15,7 +15,8 @@ import (
 // Default service addresses and URLS of Argo CD internal services
 const (
 	// DefaultRepoServerAddr is the gRPC address of the Argo CD repo server
-	DefaultRepoServerAddr = "argocd-repo-server:8081"
+	DefaultRepoServerAddr        = "argo-cd-repo-server:8081"
```
Comment (pasha-codefresh) [inline common/common.go]: revert

File: controller/sync.go [hunk 7]
```diff
@@ -524,7 +524,7 @@ func hasSharedResourceCondition(app *v1alpha1.Application) (bool, string) {
 }
 
 // delayBetweenSyncWaves is a gitops-engine SyncWaveHook which introduces an artificial delay
-// between each sync wave. We introduce an artificial delay in order give other controllers a
+// between each sync wave. We introduce an artificial delay in order give other controller a
```
Comment (pasha-codefresh) [inline controller/sync.go]: revert

File: event_reporter/server.go [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+	settingsMgr    *settings_util.SettingsManager
+	enf            *rbac.Enforcer
+	projInformer   cache.SharedIndexInformer
+	projLister     applisters.AppProjectNamespaceLister
+	policyEnforcer *rbacpolicy.RBACPolicyEnforcer
+	appInformer    cache.SharedIndexInformer
+	appLister      applisters.ApplicationLister
+	db             db.ArgoDB
+
+	// stopCh is the channel which when closed, will shutdown the Argo CD server
```
Comment (pasha-codefresh) [inline event_reporter/server.go]: rename

File: event_reporter/server.go [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+	if !cache.WaitForCacheSync(ctx.Done(), a.projInformer.HasSynced, a.appInformer.HasSynced) {
+		log.Fatal("Timed out waiting for project cache to sync")
+	}
+
+	a.stopCh = make(chan struct{})
+	<-a.stopCh
+}
+
+// NewServer returns a new instance of the Argo CD API server
```
Comment (pasha-codefresh) [inline event_reporter/server.go]: rename

File: manifests/base/event-reporter/event-reporter-service.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app.kubernetes.io/name: argocd-server
```
Comment (pasha-codefresh) [inline manifests/base/event-reporter/event-reporter-service.yaml]: rename

File: cmd/event-reporter-server/commands/event_reporter_server.go [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+	command.Flags().BoolVar(&repoServerPlaintext, "repo-server-plaintext", env.ParseBoolFromEnv("EVENT_REPORTER_REPO_SERVER_PLAINTEXT", false), "Use a plaintext client (non-TLS) to connect to repository server")
+	command.Flags().BoolVar(&repoServerStrictTLS, "repo-server-strict-tls", env.ParseBoolFromEnv("EVENT_REPORTER_REPO_SERVER_STRICT_TLS", false), "Perform strict validation of TLS certificates when connecting to repo server")
+	command.Flags().StringVar(&codefreshUrl, "codefresh-url", env.StringFromEnv("CODEFRESH_URL", "https://g.codefresh.io"), "Codefresh API url")
+	command.Flags().StringVar(&codefreshToken, "codefresh-token", env.StringFromEnv("CODEFRESH_TOKEN", ""), "Codefresh token")
+	command.Flags().StringVar(&shardingAlgorithm, "sharding-method", env.StringFromEnv(common.EnvEventReporterShardingAlgorithm, common.DefaultEventReporterShardingAlgorithm), "Enables choice of sharding method. Supported sharding methods are : [legacy] ")
```
Comment (oleksandr-codefresh) [inline cmd/event-reporter-server/commands/event_reporter_server.go:208]: are ?

File: event_reporter/metrics/metrics.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+		},
+		[]string{"reporter_shard", "metric_event_type"},
+	)
+)
+
+// NewMetricsServer returns a new prometheus server which collects api server metrics
+func NewMetricsServer(host string, port int) *MetricsServer {
+	mux := http.NewServeMux()
+	registry := prometheus.NewRegistry()
+	//registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
```
Comment (oleksandr-codefresh) [inline event_reporter/metrics/metrics.go]: do we need them?

File: event_reporter/reporter/feature_manager.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+	"time"
+)
+
+type FeatureManager struct {
+	settingsMgr *settings_util.SettingsManager
+	shouldRun   bool
+}
+
+func NewFeatureManager(settingsMgr *settings_util.SettingsManager) *FeatureManager {
+	return &FeatureManager{settingsMgr: settingsMgr}
```
Comment (oleksandr-codefresh) [inline event_reporter/reporter/feature_manager.go:15]: maybe call shouldRun here ?

File: event_reporter/sharding/sharding.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	return func(app *v1alpha1.Application) (bool, int) {
+		expectedShard := distributionFunction(app)
+		// TODO: [reporter] provide ability define label with shard number
+		return expectedShard == shard, expectedShard
+	}
+}
+
+// GetDistributionFunction returns which DistributionFunction should be used based on the passed algorithm and
+// the current datas.
+func (s *sharding) GetDistributionFunction(shardingAlgorithm string) DistributionFunction {
```
Comment (oleksandr-codefresh) [inline event_reporter/sharding/sharding.go:44]: will be here err in future ?

File: event_reporter/sharding/sharding.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	replicas := env.ParseNumFromEnv(argocommon.EnvEventReporterReplicas, 0, 0, math.MaxInt32)
+	return func(a *v1alpha1.Application) int {
+		if replicas == 0 {
+			return -1
+		}
+		if a == nil {
+			return 0
+		}
+		id := a.Name
+		log.Debugf("Calculating application shard for cluster id: %s", id)
```
Comment (oleksandr-codefresh) [inline event_reporter/sharding/sharding.go]: remove ?

File: manifests/base/event-reporter/event-reporter-statefulset.yaml [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+            configMapKeyRef:
+              name: codefresh-cm
+              key: base-url
+              optional: true
+        - name: CODEFRESH_TOKEN
+          valueFrom: 
+             secretKeyRef: 
+               key: token
+               name: codefresh-token
+        # todo: clean up
```
Comment (oleksandr-codefresh) [inline manifests/base/event-reporter/event-reporter-statefulset.yaml]: ?

File: server/application/application.go [hunk 17]
```diff
@@ -1121,8 +1121,16 @@ func (s *Server) StartEventSource(es *events.EventSource, stream events.Eventing
 	for {
 		select {
 		case event := <-eventsChannel:
+			log.Infof("event channel size is %d", len(eventsChannel))
+			rVersion, _ := s.settingsMgr.GetCodefreshReporterVersion()
+			if rVersion == string(settings.CodefreshV2ReporterVersion) {
+				logCtx.Info("v1 reported disabled skipping event")
```
Comment (oleksandr-codefresh) [inline server/application/application.go:1127]: debug lvl

File: util/settings/settings.go [hunk 18]
```diff
@@ -202,6 +202,14 @@ type KustomizeSettings struct {
 	Versions     []KustomizeVersion
 }
 
+// CodefreshReporterVersion includes all cf reporter versions
+type CodefreshReporterVersion string
+
+const (
+	CodefreshV1ReporterVersion CodefreshReporterVersion = "v1"
```
Comment (oleksandr-codefresh) [inline util/settings/settings.go:209]: so we need v1 / v2 or boolean as @yaroslav-codefresh suggested ?

File: event_reporter/sharding/sharding.go (hunk 14 above)
Comment (pasha-codefresh) [reply event_reporter/sharding/sharding.go:44]: i think no, it always will be some algorithm that is default

File: server/application/application.go (hunk 17 above)
Comment (pasha-codefresh) [reply server/application/application.go:1127]: would keep as info for now

File: util/settings/settings.go (hunk 18 above)
Comment (pasha-codefresh) [reply util/settings/settings.go:209]: not matter for now

File: manifests/base/event-reporter/event-reporter-statefulset.yaml [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+      app.kubernetes.io/name: event-reporter
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: event-reporter
+    spec:
+      serviceAccountName: event-reporter
+      containers:
+      - name: event-reporter
+        image: quay.io/argoproj/argocd:latest
```
Comment (yaroslav-codefresh) [inline manifests/base/event-reporter/event-reporter-statefulset.yaml:23]: need to provide our own

**Pull request merged**